[FEAT] admin page 생성 후 기능 작업

her0807 opened the pull request:
## 구현 내용
관리자 페이지에서 사이트를 관리 할 수 있도록 한다.
관리자만 해당 페이지에서 조작이 가능하도록 한다.
로그인은 소셜로그인을 기반으로 한다. 

1) thymeleaf + 부트 스트랩 사용해서 화면 구현
2) 플래그먼트 사용해서 조립

## 관리자 페이지 중 `회원 관리` 에서 할 수 있어야 하는 기능

- [x] 1) 회원 ` 탈퇴 ` 
- [x] 2) 회원 -> 관리자 , 관리자 -> 회원으로  `롤 변경`

## 필요한 필드
이메일, 이름, role,  

---

## 관리자 페이지 중 `게시글 관리` 에서 할 수 있어야 하는 기능

- [x] 1) 게시글 ` 삭제 ` 상태 변경 
- [x] 2) 기존에 있는 카테고리 `카테고리 변경` ( 카테고리가 추가되면 여기도 추가 되어야 함)


## 필요한 필드
글쓴이, 제목 , 조회수, 카테고리, 삭제 상태  

---

## 관리자 페이지 중 `질문 / 댓글 관리` 에서 할 수 있어야 하는 기능

- [x] 1) 질문, 댓글 비공개 `상태 변경` 
- [x] 2) 질문 별 댓글 보기 
- [ ] 3) 삭제 


## 필요한 필드
 제목 , 상태, 삭제, 댓글 보기 

---


## 관리자 페이지 중 `메인 화면` 에서 할 수 있어야 하는 기능

- [x] 1) 현재 카테고리 조회
- [x] 2) 카테고리 가중치 변경
- [x] 3) 카테고리 삭제
- [x] 4) 카테고리 추가


## 필요한 필드
 제목 , 상태, 삭제, 댓글 보기 




#66

her0807 committed:
feat: rebase 후 어드민 페이지 model - view 방식으로 구성
1) gradle 내에 있는 설정 확인
2) userapicontroller import 설정 확인

hyeonic commented on src/main/java/com/skhuedin/skhuedin/social/OauthController.java:
```diff
@@ -63,7 +66,9 @@
         token = userService.signIn(user);
 
         UserMainResponseDto responseDto = new UserMainResponseDto(userService.findByEmail(user.getEmail()));
+        MultiValueMap<String, String> header = new LinkedMultiValueMap<>();
+        header.add("Authorization", "Bearer " + token);
         // user 인증을 위한 자체 토큰을 발급받아  저장,데이터에 user 값도 저장 해서 보냄
-        return ResponseEntity.status(HttpStatus.OK).body((new TokenWithCommonResopnse(responseDto, token)));
+        return new ResponseEntity<>(new CommonResponse<>(user),header, HttpStatus.OK);
```
저희가 header 쪽에 값을 넣어줘야 하는 건가요??

hyeonic reviewed (commented):
comment 달았습니다. 확인해주세요오

hyeonic commented on src/main/java/com/skhuedin/skhuedin/interceptor/AuthorizationExtractor.java:
```diff
@@ -1,18 +1,23 @@
 
     public String extract(HttpServletRequest request, String type) {
         Enumeration<String> headers = request.getHeaders(AUTHORIZATION);
+        log.info("여기 들어옴 111111111111111111111111111111111111");
+
```
디버깅이 완료되면 삭제해주세요!

hyeonic commented on build.gradle:
```diff
@@ -37,11 +37,18 @@ dependencies {
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
 
+    implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
+    /*
+    thymeleaf -layout 을 사용하기 위해서 설정 추가 - 5/6일 수진
+     */
+    implementation 'nz.net.ultraq.thymeleaf:thymeleaf-layout-dialect'
```
사용하지 않는 것은 삭제하는 게 좋을 것 같아요!

hyeonic reviewed (commented):
comment 확인 부탁 드려요!

hyeonic commented on build.gradle:
```diff
@@ -37,11 +37,15 @@ dependencies {
             'io.jsonwebtoken:jjwt-jackson:0.11.2'
 }
 
+
+
```
불필요한 공백 제거하면 더욱 좋을 것 같아요. 그리고 상단 부분에

![image](https://user-images.githubusercontent.com/59357153/118648837-cb577180-b81d-11eb-980b-81f2ed7b27d3.png)

이미 thymeleaf가 있어요. 확인 부탁드립니다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/config/WebMvcConfig.java:
```diff
@@ -26,4 +26,5 @@ public void addCorsMappings(CorsRegistry registry) {
                 .allowedOrigins("http://localhost:3000");
     }
-}
+
+}
```
동일합니다. 공백 줄여주세요!

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,102 @@
+
+    @GetMapping("/board/main")
+    public String sign(HttpServletRequest request, Model model, HttpServletResponse response) {
+
+        Map<String, ?> inputFlashMap = RequestContextUtils.getInputFlashMap(request);
+        UserMainResponseDto dto = (UserMainResponseDto) inputFlashMap.get("user");
+//        String token = (String)inputFlashMap.get("Authorization");
+//        response.addHeader("Authorization",token);
+        // 30분후 만료되는 jwt 만들어서 쿠키에 저장
+//        Cookie cookie = new Cookie("Authorization", token);
+//        cookie.setPath("/");
+//        cookie.setMaxAge(Integer.MAX_VALUE);
+//        response.addCookie(cookie);
+
+
+        model.addAttribute("user", dto);
+        return "board/main";
+
+    }
+
+    @GetMapping("/board/userList")
+    public String findAll(HttpServletRequest request, Model model) {
+        List<UserMainResponseDto> list = userService.findAll();
+        model.addAttribute("list", list);
+        return "board/userList";
+    }
+
+    @GetMapping("/board/userList/{userId}")
+    public String userDetail(@PathVariable("userId") Long id, Model model) {
+        UserMainResponseDto user = userService.findById(id);
+        model.addAttribute("user", user);
+        return "board/userDetail";
+    }
+
+    @PostMapping("/board/delete/{userId}")
+    public String userDelete(@PathVariable("userId") Long id) {
+        userService.delete(id);
+        return "redirect:/board/userList";
+    }
+}
```
admin page를 어떻게 구상할지 확실히 정해지면 추가해도 좋을 것 같아요. 지금 당장 사용하지 않는 코드들이 함께 뒤섞여 있으면 혼동을 야기할 것 같아요!

hyeonic reviewed (commented):
확인 했습니다. 우선 어떻게 admin page를 구상하고 적용할지 고민해보는 것이 우선 인 것 같아요. 이번 PR을 단순히 반영하기에는 무리가 있을 것으로 판단됩니다. 우선 부트 스트랩 부터 적용을 하고 차근차근 화면 구성을 고민해보는 것이 어떨까요?

her0807 committed:
feat: admin page 기본 layout

her0807 committed:
feat: rest 통신 방식으로 변경/ 카테고리 가중치 로직 생성

1) js 에서 json 을 받아서 뷰를 구성하는 로직으로 변경
2) 회원,게시글,질문,카테고리  기본적인 조회 구현
3) 카테고리 가중치 up/down 구현

her0807 committed:
feat: user delete 구현

her0807 commented:
# Discussion
@hyeonic 
---

1) 게시글이 삭제되면, 아래와 같이 게시글이 삭제되는 것이 아니라, 게시글의 내용이 삭제되었다고
바뀜. 
![1](https://user-images.githubusercontent.com/26570275/119092190-143f3e00-ba49-11eb-9f4f-2d4e837a9f69.PNG)

2) USER 가 삭제되기 위해서는 그와 `연관 관계를 맺고 있는   COMMENT. BLOG, QUESTION 이 먼저 삭제` 된 후, 외래 키 설정이 되어있지 않은 유저만 삭제하도록 구현. 즉 USER 를 삭제하기 위해서는 
외래키를 제거하기 위해서 게시글 내용이 삭제되었다고 바뀌는 게 아니라` 게시글 자체가 삭제`되는 문제 발견. 

---

# 해결 방법
1) 유저 또한 삭제하지 않고, 삭제되었다는 표시만 남김. 
(그렇게 된다면 메모리 유지는 가능하겠지만, 낭비될 것 같음) 

2) 지금 처럼 삭제시킨다.

her0807 committed:
feat: 카카오로그인 구현 후 JWT TOKEN  저장 로직 생성/  role 검증 로직 수정

1) 소셜로 로그인을 하게 되면 서버에서 토큰 발급 후 뿌려줌
2) 프론트에서 잡아서 로컬 스토리지에 저장 후 권한이 필요한 일을 수행하기 전 꺼내서 header 에 보냄
3) 서버 controller 에 @MyRole 로 권한 설정
4) 불필요한 공백 + 로깅 내역 수정

hyeonic commented on build.gradle:
```diff
@@ -24,25 +24,34 @@ dependencies {
+
+
+
+    implementation group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.10.0'
```
jacskon core의 경우 이미 spring web에 내장되어 있는 걸로 아는데 따로 추가하신 이유가 있나요?

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,143 @@
+@RequiredArgsConstructor
+@Slf4j
+public class
+
+
```
공백 확인 부탁드립니다!

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,143 @@
+
+    @MyRole(role = Role.ADMIN)
+    @ResponseBody
+    @RequestMapping(value = "/userList", method = RequestMethod.POST)
```
@PostMapping으로 수정하셔도 될 것 같아요

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,143 @@
+    @MyRole(role = Role.ADMIN)
+    @ResponseBody
+    @RequestMapping(value = "/userList", method = RequestMethod.POST)
+    public List<UserMainResponseDto> userList() {
+        List<UserMainResponseDto> list = userService.findAll();
+        return list;
+    }
```
후에 user가 늘어나면 user를 검색하는 정렬조건에 기반하여 paging 하는 것도 고려해 보면 좋을 것 같습니다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,143 @@
+        return "contents/userList";
+    }
+
+    @PostMapping("/userDelete")
```
url 부분은 보통 대문자를 명시하지 않습니다. 개발환경이 window에서는 대소문자를 구분하지 않지만 linux에서는 대소문자를 구분하기 때문에 소문자로 통일하지 않으면 예기치 못한 에러를 찾기 어려울 수 있어요!

delete를 표기하고 싶다면 /user-delete 와 같이 명시하는 것이 좋을 것 같습니다.

her0807 replied:
앗 그렇군요.. 당장만 생각을 했던 것 같습니다. 조금 더 url 고민 후 재작성 하겠습니다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,143 @@
+        return "contents/categoryList";
+    }
+
+    @ResponseBody
+    @PostMapping("/questionDetail")
+    public List<CommentMainResponseDto> commentMainList(@RequestParam(value = "id", required = false, defaultValue = "0") Long id) {
+        return  commentService.findByQuestionId(id);
+    }
```
이렇게 comment와 관련된 api는 CommentApiController에 추가하는 것이 어떨까요?

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,143 @@
+        postsService.deleteAdmin(id);
+        return "redirect:/postList";
+    }
+    @MyRole(role = Role.ADMIN)
+    @ResponseBody
+    @RequestMapping(value = "/postList", method = RequestMethod.POST)
+    public List<PostsAdminResponseDto> postList() {
+        List<PostsAdminResponseDto> list = postsService.findAll();
+        return list;
+    }
```
Posts와 관련된 api이기 때문에 postsApiController에 위치하는 것은 어떨까요? Mapping method 또한 @PostMapping으로 짧게 줄이는 것이 더욱 바람직할 것 같아요!

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,143 @@
+        return "redirect:/postList";
+    }
+
+    @PostMapping("categoryList/up")
+    public String upCategory(@RequestParam(value = "id", required = false, defaultValue = "0") Long id) {
+        categoryService.addWeight(id);
+        return "redirect:/categoryList";
+    }
+
+    @PostMapping("categoryList/down")
+    public String downCategory(@RequestParam(value = "id", required = false, defaultValue = "0") Long id) {
+        categoryService.subtractWeight(id);
+        return "redirect:/categoryList";
+    }
+
+    @PostMapping("postDelete")
+    public String deleteCategory(@RequestParam(value = "id", required = false, defaultValue = "0") Long id) {
+        postsService.deleteAdmin(id);
+        return "redirect:/postList";
+    }
```
이렇게 단순히 데이터를 조작하기 위한 핸들링 메소드들은 따로 권한을 부여하지 않아도 괜찮을까요?

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,143 @@
+    public String admin() {
+        return "board/login";
+    }
+
+
+
+
```
줄바꿈 제거 부탁드립니다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/domain/Posts.java:
```diff
@@ -56,6 +56,13 @@ public void updatePost(Posts post) {
         this.category = post.category;
     }
 
+    public void updateCategory(Category category){
+        this.category =category;
+    }
+    public void updateContent(){
+        this.content ="관리자에 의해 삭제되었습니다.";
+    }
```
항상 commit 전에 패키지 단에서 ctrl + alt + l을 통하여 프로젝트의 서식을 통일하여 위와 같이 미쳐 발견하지 못한 공백 부분이 정리될 수 있어요! 꼭 적용 부탁드립니다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/dto/category/CategoryMainResponseDto.java:
```diff
@@ -0,0 +1,20 @@
+        this.name = category.getName();
+        this.weight = category.getWeight();
+    }
+
+
```
동일합니다ㅏ 줄바꿈 제거 부탁드려요

hyeonic commented on src/main/java/com/skhuedin/skhuedin/dto/user/AdminSaveRequestDto.java:
```diff
@@ -0,0 +1,34 @@
+package com.skhuedin.skhuedin.dto.user;
+
+import com.skhuedin.skhuedin.domain.User;
+import com.skhuedin.skhuedin.infra.Role;
+import lombok.Builder;
+
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@NoArgsConstructor
+public class AdminSaveRequestDto {
+
+    private String email;
+    private String password;
+
+    @Builder
+    public AdminSaveRequestDto(String email,
+                               String password) {
+        this.email = email;
+        this.password = password;
+    }
+
+    public User toEntity() {
+        User user = User.builder()
+                .email(this.email)
+                .password(this.password)
+                .role(Role.ADMIN)
+                .build();
+        return user;
+    }
+}
```
소셜 도입 전 사용하기 위한 용도 인가요?

her0807 replied:
네! 소셜 도입 전, 사용했던 로그인 로직을 위해서 작성했던 코드인데, 지금은 필요없어졌습니다. 삭제하겠습니다!

hyeonic commented on src/main/java/com/skhuedin/skhuedin/interceptor/AuthorizationExtractor.java:
```diff
@@ -1,19 +1,24 @@
+
         while (headers.hasMoreElements()) {
             String value = headers.nextElement();
+            log.info(value);
```
logging이 완료되었으면 삭제하셔도 좋을 것 같아요!

hyeonic commented on src/main/java/com/skhuedin/skhuedin/service/UserService.java:
```diff
@@ -38,6 +55,48 @@ public void update(Long id, UserUpdateDto updateDto) {
+        List<Posts> posts;
+
+
+        if (comments != null) {
+            for (Comment comment : comments) {
+                commentRepository.delete(comment);
+            }
+        }
+
+        if (questions != null) {
+            for (Question question : questions) {
+                List<Comment> innerComment = commentRepository.findByQuestionId(question.getId());
+                if (comments != null) {
+                    for (Comment comment : innerComment) {
+                        commentRepository.delete(comment);
+                    }
+                }
+                questionRepository.delete(question);
+            }
+        }
+
+        if (targetQuestions != null) {
+            for (Question question : targetQuestions) {
+                questionRepository.delete(question);
+            }
+        }
+
+        if (blogByUserId != null) {
+            posts = postsRepository.findPostsByBlogId(blogByUserId.getId());
+
+            for (Posts post : posts) {
+                log.info(post.getId().toString());
+                postsRepository.deleteById(post.getId());
+            }
+            blogRepository.delete(blogByUserId);
+        }
```
전부 null로 처리를 하셨는데 보통 예외를 던지도록 메소드를 작성하지 않으셨나요? 그렇게 되면 null 값은 될 수 없고 throw된 예외를 catch해야 할 것 같은데 한번 확인 부탁드려요!

her0807 replied:
![1](https://user-images.githubusercontent.com/26570275/119282758-ca876b00-bc75-11eb-97a6-4ec3117b9424.PNG)

현재 UserRepository 에만 Optional 처리가 되어 있어서 그부분은  findUser 와 같이 단일 객체일 경우 optional 에서 객체가 무조건 get 되어야 할 경우에 null 이면 에러가 나기 때문에 전처리를 해주었습니다. 

그런데 위에 있는 comments, questions, targetQuestions  와 같이 배열 형태로 넘어온 객체들은 조회 시 값이 없으면 null 상태이기 때문에 null로 값을 겸증하는 로직을 작성했었습니다!

hyeonic replied:
List로 목록 조회할 경우 **빈 List 객체**를 반환하는 것으로 알고 있어요! 즉 아무것도 가리키지 않는 null을 return 하는 것이 아니기 때문에 해당 List 객체의 isEmpty() 메소드를 활용해야 하지 않을까요? 확인 부탁드립니당

her0807 replied:
@hyeonic  앗 안그래도 널로 바로 접근하는 것보다  isEmpty() 메소드로 바꾸는게 더 좋을 것 같아서 생각만 하고 있었는데
말씀해주셔서 감사합니다!!

hyeonic commented on src/main/java/com/skhuedin/skhuedin/social/OauthController.java:
```diff
@@ -52,11 +54,14 @@
 
         Long id = null;
         // 사용자가 현재 회원인지 아닌지 확인 작업. 회원이 아니면 회원 가입을 시키고
+        log.info(user.getEmail());
         if (userService.findByEmail(user.getEmail()) == null) {
+            log.info(user.getEmail()+ "여기 들어오는지?");
             userService.signUp(user);
         }
         //회원이면 로그인을 시킴
         token = userService.signIn(user);
+        log.info(token);
```
확인이 끝나면 지워주세요!

hyeonic reviewed (commented):
comment 확인 부탁드립니다!

her0807 committed:
feat: category 생성/ CSS 작업

1) 카테고리 생성 작업 구현
2) 관련 페이지 css 작업

her0807 commented:
# 카테고리 삭제

카테고리를 삭제하게 되면, 참조하고 있는 글이 있는 경우  외래키로 사용하고 있어 삭제할 수 없다. 

## 해결방안
1) 카테고리를 삭제하지말고 상태로 처리
2) 참고하고 있는 포스트의 수를 보여주고, 포스트의 카테고리로 등록되어 삭제하지 못한다는 에러 창 띄움

her0807 committed:
feat: category delete
1) category delete 구현
post service 에 findbycategoryid 만들어서
카테고리를 참조하고 있는 글의 갯수를 파악
참조하고 있는 글이 없는 경우 삭제가 가능하게 구현

her0807 commented:
# 관리자 페이지 중 `회원 관리` 에서 할 수 있어야 하는 기능

1) 회원 ` 탈퇴 `
2) 회원 -> 관리자 , 관리자 -> 회원으로  `롤 변경`

## 필요한 필드
이메일, 이름, role,

her0807 commented:
# 관리자 페이지 중 `게시글 관리` 에서 할 수 있어야 하는 기능

1) 게시글 ` 삭제 ` 상태 변경 
2) 기존에 있는 카테고리 `카테고리 변경`

## 필요한 필드
글쓴이, 제목 , 조회수, 카테고리, 삭제 상태

her0807 committed:
feat: category 가 추가되면 post 에 카테고리 선택에 추가되도록 변경
기존에는 <option> 값을 고정해뒀는데, category 가 추가되면
변경이 필요하기 때문에 함수를 만들어 추가했음

her0807 commented:
## post 에서 카테고리를 불러오는 이슈에서 리턴 값이 undefined 로 되는 이슈가 있었음
이건 AJAX 에서 비동기 방식을 제공하기 때문임 

### 동기 방식
쉽게 이야기하면, 하나의 함수가 끝나고 반환 값을 받은 뒤 다음 함수를 실행하는 것.

### 비동기 방식
함수가 끝나기 전에 다른 함수를 호출하는 것 
---
### 참고 자료

https://blog.silnex.kr/ajax%EB%A5%BC-%ED%86%B5%ED%95%B4-%EB%B3%B4%EB%8A%94-%EB%8F%99%EA%B8%B0-%EB%B0%A9%EC%8B%9D%EA%B3%BC-%EB%B9%84%EB%8F%99%EA%B8%B0-%EB%B0%A9%EC%8B%9D%EC%9D%98-%EC%9D%B4%ED%95%B4/

https://tjddnjs625.tistory.com/5

her0807 commented:
# 관리자 페이지 중 `질문 / 댓글 관리` 에서 할 수 있어야 하는 기능

1) 질문, 댓글 비공개 `상태 변경` 
2) 질문 별 댓글 보기 
3) 삭제 

## 필요한 필드
 제목 , 상태, 삭제, 댓글 보기

her0807 committed:
feat: login view 수정 + 로그인 전 화면 설정
1) 로그인을 하지 않았을 경우, 보여지는 화면을 구성한다.

her0807 commented on src/main/resources/templates/contents/kakaoLogin.html:
```diff
@@ -0,0 +1,36 @@
+
+</head>
+<body>
+<div class="contents" th:fragment="kakaoLogin">
+    <div class="category">
+        <div class="container">
+            <br/>
+            <h2>[관리자 페이지 로그인]</h2>
+            <br/>
+            <h4>SKHUED-IN 페이지를 위한 공간! </h4>
+            <h4>관리자 아닌 분들은 접근이 불가능합니다 :)</h4>
+            <br/>
+            <a id="kakao-login-btn"></a>
+            <a href="http://developers.kakao.com/logout"></a>
+            <script type="text/javascript">
+
+                Kakao.init('b9de844ef665dda98917b7ec7b9d9a8b');
+                Kakao.Auth.createLoginButton({
+                    container: '#kakao-login-btn',
+                    success: function (authObj) {
+                        kakao(authObj["access_token"]);
+                    },
+                    fail: function (err) {
+                        alert(JSON.stringify(err));
+                    }
+                });
+            </script>
+        </div>
+    </div>
```
## 로그인 화면

## 설명
카카오 소셜 로그인과 연동하여 엑세스 토큰을 발급 받고, 토큰을 서버로 넘겨주어 로그인을 수행한다. 

![image](https://user-images.githubusercontent.com/26570275/119944074-cc7f6000-bfce-11eb-8fe3-ad043461ca57.png)

her0807 commented on src/main/resources/static/js/get.js:
```diff
@@ -0,0 +1,366 @@
+            return $('#result').html(result)
+        }
+    })
+   return $('#result').html(defaultText())
+}
```
## 로그인 전 화면

## 내용
로그인이 안된 사용자에게 보여지는 기본 화면

![image](https://user-images.githubusercontent.com/26570275/119944484-49123e80-bfcf-11eb-835f-33a5787405f3.png)

her0807 commented on src/main/resources/static/js/get.js:
```diff
@@ -0,0 +1,366 @@
+                result += '<td><button type="button" onclick=userDelete(' + item["id"] + ')>삭제</button></td></tr>'
+            })
+            result += '</table>'
+            $('#result').html(result)
```
## 회원 조회 

## 내용

회원을 조회 하면 보여지는 화면 구성

![image](https://user-images.githubusercontent.com/26570275/119944721-955d7e80-bfcf-11eb-8931-523d259f571a.png)

her0807 commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,158 @@
+        return "redirect:/categoryList";
+    }
+
+    @PostMapping("user/role")
+    public String userRoleChange(@RequestParam(value = "id", required = false, defaultValue = "0") Long id) {
+        userService.updateRole(id);
+        return "redirect:/userList";
+    }
```
## 유저 권한 변경

## 내용
유저 -> 관리자 
관리자->유저
로 변경한다. 
![image](https://user-images.githubusercontent.com/26570275/119945025-ee2d1700-bfcf-11eb-96e7-32096ef59523.png)

her0807 commented on src/main/java/com/skhuedin/skhuedin/service/UserService.java:
```diff
@@ -34,10 +51,60 @@ public void update(Long id, UserUpdateDto updateDto) {
+        List<Question> questions = questionRepository.findQuestionByUserId(id);
+        List<Question> targetQuestions = questionRepository.findQuestionByTargetUserId(id);
+        List<Posts> posts;
+
+        if (!comments.isEmpty()) {
+            for (Comment comment : comments) {
+                commentRepository.delete(comment);
+            }
+        }
+
+        if (!questions.isEmpty()) {
+            for (Question question : questions) {
+                List<Comment> innerComment = commentRepository.findByQuestionId(question.getId());
+                if (comments != null) {
+                    for (Comment comment : innerComment) {
+                        commentRepository.delete(comment);
+                    }
+                }
+                questionRepository.delete(question);
+            }
+        }
+
+        if (!targetQuestions.isEmpty()) {
+            for (Question question : targetQuestions) {
+                questionRepository.delete(question);
+            }
+        }
+
+        if (blogByUserId != null) {
+            posts = postsRepository.findPostsByBlogId(blogByUserId.getId());
+
+            for (Posts post : posts) {
+                log.info(post.getId().toString());
+                postsRepository.deleteById(post.getId());
+            }
+            blogRepository.delete(blogByUserId);
+        }
         userRepository.delete(findUser);
     }
 
```
## 회원 삭제 

## 내용
유저에 관련된 연관 관계를 모두 삭제하고, 유저를 삭제한다. 

![image](https://user-images.githubusercontent.com/26570275/119945572-97740d00-bfd0-11eb-9855-7cac1b351175.png)

her0807 commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,158 @@
+        return "redirect:/postList";
+    }
+
+    @MyRole(role = Role.ADMIN)
+    @ResponseBody
+    @PostMapping("/postList")
+    public List<PostsAdminResponseDto> postList() {
+        List<PostsAdminResponseDto> list = postsService.findAll();
+        return list;
+    }
```
## 게시글 조회
## 내용
게시글에 관련된 전반적인 것들을 조회한다. 관리자만 가능하게한다. 

![image](https://user-images.githubusercontent.com/26570275/119945799-e457e380-bfd0-11eb-8374-c5b5c4d03ac7.png)

her0807 commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,158 @@
+        return commentService.findByQuestionId(id);
+    }
+
+    @PostMapping("update/category")
+    public String updateCategory(@RequestParam(value = "category", required = false, defaultValue = "0") Long category,
+                                 @RequestParam(value = "post_id", required = false, defaultValue = "0") Long post_id) {
+        postsService.update(post_id, category);
+        return "redirect:/postList";
+    }
```
## 게시글의 카테고리를 업데이트

![image](https://user-images.githubusercontent.com/26570275/119945993-123d2800-bfd1-11eb-8899-07e54d53a38d.png)

her0807 commented on src/main/java/com/skhuedin/skhuedin/service/PostsService.java:
```diff
@@ -38,17 +43,48 @@ public Long update(Long id, PostsSaveRequestDto requestDto) {
         postsRepository.delete(posts);
     }
 
+
+    @Transactional
+    public void deletePostAdmin(Long id) {
+        Posts posts = getPosts(id);
+        posts.setDeleteStatus();
+    }
+
```
## 게시글 삭제

## 내용
게시글의 상태를 삭제로 변경한다. 

![image](https://user-images.githubusercontent.com/26570275/119946257-65af7600-bfd1-11eb-9c90-c272704d1abf.png)

her0807 commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    @MyRole(role = Role.ADMIN)
+    @ResponseBody
```
## 질문/ 댓글 조회

## 내용
질문 댓글을 조회한다. 

![image](https://user-images.githubusercontent.com/26570275/119946428-94c5e780-bfd1-11eb-85f7-a425a9ecf9cd.png)

her0807 commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,158 @@
+        return "contents/createCategory";
+    }
+
+    @ResponseBody
+    @PostMapping("/questionDetail")
+    public List<CommentMainResponseDto> commentMainList(@RequestParam(value = "id", required = false, defaultValue = "0") Long id) {
+        return commentService.findByQuestionId(id);
+    }
+
```
## 질문 별 댓글
 
![image](https://user-images.githubusercontent.com/26570275/119946631-cd65c100-bfd1-11eb-9a04-539fda30f080.png)

her0807 commented on src/main/java/com/skhuedin/skhuedin/service/QuestionService.java:
```diff
@@ -43,6 +46,17 @@ public Long update(Long id, QuestionSaveRequestDto requestDto) {
         return question.getId();
     }
 
+    @Transactional
+    public void updateStatus(Long id) {
+        Question question = getQuestion(id);
+
+        if (question.getStatus() == false) {
+            question.updateStatus(true);
+        } else if (question.getStatus() == true) {
+            question.updateStatus(false);
+        }
+    }
+
```
## 질문 비공개

her0807 commented on src/main/java/com/skhuedin/skhuedin/controller/AdminController.java:
```diff
@@ -0,0 +1,158 @@
+        return "redirect:/postList";
+    }
+
+    @PostMapping("categoryList/up")
+    public String upCategory(@RequestParam(value = "id", required = false, defaultValue = "0") Long id) {
+        categoryService.addWeight(id);
+        return "redirect:/categoryList";
+    }
+
+    @PostMapping("categoryList/down")
+    public String downCategory(@RequestParam(value = "id", required = false, defaultValue = "0") Long id) {
+        categoryService.subtractWeight(id);
+        return "redirect:/categoryList";
+    }
+
+    @PostMapping("create/category")
+    public String createCategory(@RequestBody CategoryRequestDto categoryRequestDto) {
+        categoryService.save(categoryRequestDto);
+        return "redirect:/categoryList";
+    }
```
## 카테고리 생성 + 가즁치 변경

![image](https://user-images.githubusercontent.com/26570275/119946938-1e75b500-bfd2-11eb-9a9f-731b513264cc.png)

her0807 reviewed (commented):
작업 내용 간단히 정리했습니다!

hyeonic commented:
전체적으로 확인하였습니다!! 두 가지 제안사항이 있습니다. 
1. 카테고리의 경우 id 값을 보여주는 것이 아니라 카테고리 이름 자체를 보여주도록 하는게 어떨까요?
2. 혹시 폰트 교체가 가능할까요?? 직관성이 너무 떨어져 보이는 것 같아요 ㅜㅠ 저희 main 화면과 맞추어 하는 것도 좋을 것 같습니다.

her0807 committed:
feat: 게시글 카테고리 선택 반영
1) 게시글 별 카테고리를 select 문에 표시 되도록 변경
2) 글꼴 변경
